modules/library/render/src/main/java/org/geotools/renderer/lite/MemoryFilterOptimizer.java

```diff
@@ -111,7 +117,58 @@ public Object visit(Not filter, Object extraData) {
 
     @Override
     public Object visit(Or filter, Object extraData) {
-        return memoize(filter, extraData, super::visit);
+        return memoize(filter, extraData, this::inFilterOptimizer);
+    }
+
+    /** Checks if an Or filter can be replaced by a single IN condition. */
+    private Object inFilterOptimizer(Or filter, Object extraData) {
+        InFunction inFilter = replaceWithInFilter(filter);
+        if (inFilter == null) return super.visit(filter, extraData);
+        return ff.equals(inFilter, ff.literal(true));
+    }
+
+    private InFunction replaceWithInFilter(Or filter) {
+        List<Filter> children = filter.getChildren();
+        if (children.size() == 0
+                || !children.stream().allMatch(f -> f instanceof PropertyIsEqualTo)) return null;
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        List<PropertyIsEqualTo> equalsList = (List) children;
+        // get the parameters pairs, including nulls when the requirements are not accomplished
+        List<Pair<PropertyName, Literal>> parameters =
+                equalsList.stream().map(eq -> getEqualsParameters(eq)).collect(Collectors.toList());
+        // check if any pair is null
+        if (parameters.stream().anyMatch(Objects::isNull)) return null;
+        // check all the property names are the same
+        String propertyName = parameters.get(0).getKey().getPropertyName();
+        if (!parameters
+                .stream()
+                .allMatch(pair -> Objects.equals(propertyName, pair.getKey().getPropertyName())))
+            return null;
+        // build the In function
+        List<Expression> inParameters = new ArrayList<>(parameters.size() + 1);
+        inParameters.add(parameters.get(0).getKey());
+        inParameters.addAll(
+                parameters.stream().map(p -> p.getValue()).collect(Collectors.toList()));
+        InFunction inFunction = new InFunction();
+        inFunction.setParameters(inParameters);
+        return inFunction;
+    }
+
+    private Pair<PropertyName, Literal> getEqualsParameters(PropertyIsEqualTo equals) {
```

Comment:
Nitpicking here, an "in" optimization can be performed over pairs of "Expression, Literal", e.g., ``strLow(myAtt) = 'abc' or strLow(myAtt) = 'def'`` or ``(a + b) = 10 or (a + 10) = 12`` can be both optimized with the in function. The only requirement is that all non literal bits in the comparison are equal. Again, probably not the most common case, up to you if you want to generalize or not.

Reply:
Nice hint, added expression support.